mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java

```diff
@@ -0,0 +1,177 @@
+            .value(BlobParts.DATA, bindMarker(BlobParts.DATA)));
+    }
+
+    public CompletableFuture<Optional<BlobId>> save(byte[] data) {
+        if (data == null) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+        BlobId blobId = BlobId.forPayload(data);
+        return saveBlobParts(data, blobId)
+            .thenCompose(partIds -> saveBlobPartsReferences(blobId, partIds))
+            .thenApply(any -> Optional.of(blobId));
+    }
+
+    private CompletableFuture<Stream<Pair<Integer, PartId>>> saveBlobParts(byte[] data, BlobId blobId) {
+        return FluentFutureStream.of(
+            dataChunker.chunk(data, CHUNK_SIZE)
+                .map(pair -> writePart(pair.getRight(), blobId, pair.getKey())
+                    .thenApply(partId -> Pair.of(pair.getKey(), partId))))
+            .completableFuture();
+    }
+
+    private CompletableFuture<PartId> writePart(ByteBuffer data, BlobId blobId, int position) {
+        PartId partId = PartId.create(blobId, position);
+        return cassandraAsyncExecutor.executeVoid(
+            insertPart.bind()
+                .setString(BlobParts.ID, partId.getId())
+                .setBytes(BlobParts.DATA, data))
+            .thenApply(any -> partId);
+    }
+
+    private CompletableFuture<Stream<Void>> saveBlobPartsReferences(BlobId blobId, Stream<Pair<Integer, PartId>> stream) {
+        return FluentFutureStream.of(stream.map(pair ->
+            cassandraAsyncExecutor.executeVoid(insert.bind()
+                .setString(Blobs.ID, blobId.getId())
+                .setLong(Blobs.POSITION, pair.getKey())
+                .setString(Blobs.PART, pair.getValue().getId()))))
+            .completableFuture();
+    }
+
+    public CompletableFuture<byte[]> read(BlobId blobId) {
+        return cassandraAsyncExecutor.execute(
+            select.bind()
+                .setString(Blobs.ID, blobId.getId()))
+            .thenApply(this::toPartIds)
+            .thenCompose(this::toDataParts)
+            .thenApply(this::concatenateDataParts);
+    }
+
+    private ImmutableMap<Long, PartId> toPartIds(ResultSet resultSet) {
+        return CassandraUtils.convertToStream(resultSet)
+            .map(row -> Pair.of(row.getLong(Blobs.POSITION), PartId.from(row.getString(Blobs.PART))))
+            .collect(Guavate.toImmutableMap(Pair::getKey, Pair::getValue));
+    }
+
+    private CompletableFuture<Stream<Optional<Row>>> toDataParts(ImmutableMap<Long, PartId> positionToIds) {
+        return CompletableFutureUtil.chainAll(
+            positionToIds.values().stream(),
+            this::readPart);
+    }
+
+    private byte[] concatenateDataParts(Stream<Optional<Row>> rows) {
+        ImmutableList<byte[]> parts = rows.flatMap(OptionalConverter::toStream)
+            .map(this::rowToData)
+            .collect(Guavate.toImmutableList());
+
+        return Bytes.concat(parts.toArray(new byte[parts.size()][]));
```

Comment:
A futur optimization would be to collect directly in a array of byte[].
But it's probably to much work for not a lot of memory wine